docs/example_nodejs.md

```diff
@@ -1,34 +1,314 @@
+</figure>
+
+Hooray, we're now using the cache instead of running `yarn install` for every single build. (For more on the principles at work here, [check out this guide](http://bitjudo.com/blog/2014/03/13/building-efficient-dockerfiles-node-dot-js/).)
+
+Here's what our timing looks like now:
+
+| Approach | Deploy Time (after initial)
+|---|---|
+| Naive | 11.31-14.21s |
+| Optimized Dockerfile | 3.25-4.12s |
+{:.benchmark-report}
+
+Pretty good! But Tilt has some tricks up its sleeve to make it even faster.
+
+## Step 3: Let's Optimize It _Even More_
+
+When we make a change to a file, we currently have to build an image, deploy new Kubernetes configs,
+and wait for Kubernetes to schedule the pod.
+
+With Tilt, we can skip all of these steps, and instead
+[live_update](live_update_tutorial.html) the pod in place.
+
+The first thing we need to do is change how our app is invoked: we're going to run it via [nodemon](https://nodemon.io/), a utility that monitors source files for changes and restarts your app as necessary.  In this branch, as reflected in [package.json](https://github.com/windmilleng/tilt-example-nodejs/blob/master/3-recommended/package.json), we've already run `yarn add --dev nodemon` to add nodemon as a dev dependency.
+
+[Our new Tiltfile](https://github.com/windmilleng/tilt-example-nodejs/blob/master/3-recommended/Tiltfile) contains the following new code:
+
+```python
+# Add a live_update rule to our docker_build
+congrats = "🎉 Congrats, you ran a live_update! 🎉"
+docker_build('example-nodejs-image', '.',
+    build_args={'node_env': 'development'},
+    entrypoint='yarn run nodemon /app/index.js',
+    live_update=[
+        sync('.', '/app'),
+        run('cd /app && yarn install', trigger=['./package.json', './yarn.lock']),
+
+        # if all that changed was start-time.txt, make sure the server
+        # reloads so that it will reflect the new startup time
+        run('touch /app/index.js', trigger='./start-time.txt'),
+
+        # add a congrats message!
+        run('sed -i "s/Hello cats!/{}/g" /app/views/index.mustache'.
+            format(congrats)),
+])
+```
+We've added some new parameters to `docker_build` that tell the container to use nodemon:
+1. the `entrypoint` parameter overrides the `ENTRYPOINT` specified in the Dockerfile; now when the container executes, it will run `yarn run nodemon /app/index.js`
+2. the `build_args` parameter corresponds to this Dockerfile change:
+    ```
+    # Default value; will be overridden by build_args, if passed
+    ARG node_env=production
+
+    ENV NODE_ENV $node_env
+    ```
+   Together, these changes mean that when we build this Dockerfile via this Tiltfile, we set the env var `$NODE_ENV=development`, and our `yarn install` call will install dev dependencies.
+
+With the help of Tilt, we've done a pretty cool thing: we've set it up so this Dockerfile can run the app for development (when invoked via Tilt) or for production (when invoked otherwise).
```

Comment:
I'm not sure this paragraph is necessary. If I'm just trying to set up my app, I don't think this would be an important or notable thing I would care about.

Reply:
Sure, but if we're trying to teach _best practices_ then I think this IS a thing we want to highlight -- we're pretty opinionated that if possible, you should use t same dockerfile for prod and dev, and just use Tilt to change the dev-specific bits, right?

Reply:
I don't want to make a bigger deal with it than this particular edit merits. 

But your comment here gets a couple things conflated/confused that we should talk about more in detail. Because they're good questions! And worth talking about!

1) The terms "opinionated tools" and "best practices" tend to get thrown around a lot in the tech community, without much introspection about what, exactly, it's worth being opinionated about and what assumptions those opinions are based on.

2) I don't think the opinion "you should use the same Dockerfile for dev and prod AT ALL COSTS" is a particularly good opinion or best practice

3) If we were going to settle on a best practice, it would be something like "prefer layers over branches" NOT "use the same dockerfile at all costs." And I'd recommend we add docker_build(env) and remove the build_args here. That tends to be more what I see in the wild.

4) The approach this example takes ("use build_args to parameterize for dev or prod") is a very narrow instance. The instant jump from specific example to broad principle comes off as very Hacker News comment-y in tone (i.e., "this worked for my codebase, so it should work for all codebases!"). 

5) There's a big gap between using a best practice in the example, and trying to explain that best practice effectively. I think trying to explain a best practice in an off-hand comment mid-guide isn't going to convey what you want.

Sorry, I know that's a lot to digest, but I hope it's helpful!

Reply:
hmmmm okay that makes sense, and i appreciate that framing. I'll cut it.

Reply:
hmmmm okay that makes sense, and i appreciate that framing. I'll cut it.